node/service/src/rpc/frequency_rpc.rs

```diff
@@ -11,43 +11,130 @@ use common_primitives::rpc::RpcEvent;
 }
 
 /// The client handler for the API used by Frequency Service RPC with `jsonrpsee`
-pub struct FrequencyRpcHandler<C, M> {
+pub struct FrequencyRpcHandler<P: TransactionPool, C, M> {
 	client: Arc<C>,
+	pool: Arc<P>,
 	_marker: std::marker::PhantomData<M>,
 }
 
-impl<C, M> FrequencyRpcHandler<C, M> {
+impl<P: TransactionPool, C, M> FrequencyRpcHandler<P, C, M> {
 	/// Create new instance with the given reference to the client.
-	pub fn new(client: Arc<C>) -> Self {
-		Self { client, _marker: Default::default() }
+	pub fn new(client: Arc<C>, pool: Arc<P>) -> Self {
+		Self { client, pool, _marker: Default::default() }
 	}
 }
 
 #[async_trait]
-impl<C, Block> FrequencyRpcApiServer<Block> for FrequencyRpcHandler<C, Block>
+impl<P, C, Block, AccountId, Nonce> FrequencyRpcApiServer<Block, AccountId, Nonce>
+	for FrequencyRpcHandler<P, C, Block>
 where
 	Block: BlockT,
+	C: HeaderBackend<Block>,
 	C: Send + Sync + 'static,
 	C: ProvideRuntimeApi<Block>,
 	C::Api: AdditionalRuntimeApi<Block>,
+	C::Api: AccountNonceApi<Block, AccountId, Nonce>,
+	P: TransactionPool + 'static,
+	AccountId: Clone + Codec,
+	Nonce: Clone + Encode + Decode + AtLeast32Bit + 'static,
 {
 	fn get_events(&self, at: <Block as BlockT>::Hash) -> RpcResult<Vec<RpcEvent>> {
 		let api = self.client.runtime_api();
 		return map_rpc_result(api.get_events(at))
 	}
+
+	fn get_missing_nonce_values(&self, account: AccountId) -> RpcResult<Vec<Nonce>> {
+		let api = self.client.runtime_api();
+		let best = self.client.info().best_hash;
+
+		let nonce = api.account_nonce(best, account.clone()).map_err(|e| {
+			CallError::Custom(ErrorObject::owned(1, "Unable to query nonce.", Some(e.to_string())))
+		})?;
+		Ok(get_missing_nonces(&*self.pool, account, nonce))
+	}
+}
+
+/// Finds any missing nonce values inside Future pool and return them as result
+fn get_missing_nonces<P, AccountId, Nonce>(pool: &P, account: AccountId, nonce: Nonce) -> Vec<Nonce>
+where
+	P: TransactionPool,
+	AccountId: Clone + Encode,
+	Nonce: Clone + Encode + Decode + AtLeast32Bit + 'static,
+{
+	// Now we need to query the transaction pool
+	// and find transactions originating from the same sender.
+	// Since extrinsics are opaque to us, we look for them using
+	// `provides` tag. And increment the nonce if we find a transaction
+	// that matches the current one.
+	let mut current_nonce = nonce.clone();
+	let encoded_account = account.clone().encode();
+	let mut current_tag = (account.clone(), nonce).encode();
+	for tx in pool.ready() {
+		// since transactions in `ready()` need to be ordered by nonce
+		// it's fine to continue with current iterator.
+		if tx.provides().get(0) == Some(&current_tag) {
+			current_nonce += One::one();
+			current_tag = (account.clone(), current_nonce.clone()).encode();
+		}
+	}
+
+	let mut result = vec![];
+	let mut my_in_future: Vec<_> = pool
+		.futures()
+		.into_iter()
+		.filter_map(|x| match x.provides().get(0) {
```

Comment:
I think rayon can be really useful here `par_iter`

Reply:
I tried that but looked more challenging than expected and got the following error
`doesn't satisfy _: IntoParallelRefIterator<'_>`

Anyway since the logic inside the iterator is simple enough I'm not too worried about the need to optimize at this point.